lib/auth/join_azure.go

```diff
@@ -0,0 +1,361 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/arm"
+	"github.com/coreos/go-oidc"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"go.mozilla.org/pkcs7"
+	"golang.org/x/exp/slices"
+	"gopkg.in/square/go-jose.v2/jwt"
+
+	"github.com/gravitational/teleport/api/client"
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/cloud/azure"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+const azureAccessTokenAudience = "https://management.azure.com/"
+
+type signedAttestedData struct {
+	Encoding  string `json:"encoding"`
+	Signature string `json:"signature"`
+}
+
+type plan struct {
+	Name      string `json:"name"`
+	Product   string `json:"product"`
+	Publisher string `json:"publisher"`
+}
+
+type timestamp struct {
+	CreatedOn string `json:"createdOn"`
+	ExpiresOn string `json:"expiresOn"`
+}
+
+type attestedData struct {
+	LicenseType    string    `json:"licenseType"`
+	Nonce          string    `json:"nonce"`
+	Plan           plan      `json:"plan"`
+	Timestamp      timestamp `json:"timestamp"`
+	ID             string    `json:"vmId"`
+	SubscriptionID string    `json:"subscriptionId"`
+	SKU            string    `json:"sku"`
+}
+
+type accessTokenClaims struct {
+	jwt.Claims
+	ResourceID string `json:"xms_mirid"`
+	TenantID   string `json:"tid"`
+	Version    string `json:"ver"`
+}
+
+type azureVerifyTokenFunc func(ctx context.Context, rawIDToken string) (*accessTokenClaims, error)
+
+type azureRegisterConfig struct {
+	clock    clockwork.Clock
+	certs    []*x509.Certificate
+	verify   azureVerifyTokenFunc
+	vmClient azure.VirtualMachinesClient
+}
+
+func azureVerifyFuncFromOIDCVerifier(cfg *oidc.Config) azureVerifyTokenFunc {
+	return func(ctx context.Context, rawIDToken string) (*accessTokenClaims, error) {
+		token, err := jwt.ParseSigned(rawIDToken)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		var unverifiedClaims accessTokenClaims
+		if err := token.UnsafeClaimsWithoutVerification(&unverifiedClaims); err != nil {
```

Comment:
Why are we not verifying the claims?

Reply:
We can't verify the claims without knowing the issuer, and the tenant ID from the claims is part of the issuer. We don't use the unverified claims for anything else.

Reply:
Maybe just add an inline comment clarifying the why